datalad/distribution/create_publication_target_sshwebserver.py

```diff
@@ -0,0 +1,270 @@
+        from os import geteuid  # Linux specific import
+        var_run_user_datalad = "/var/run/user/%s/datalad" % geteuid()
+        assure_dir(var_run_user_datalad)
+        control_path = "%s/%s" % (var_run_user_datalad, host_name)
+        control_path += ":%s" % parsed_target.port if parsed_target.port else ""
+
+        # - start control master:
+        cmd = "ssh -o ControlMaster=yes -o \"ControlPath=%s\" " \
+              "-o ControlPersist=yes %s exit" % (control_path, host_name)
+        lgr.debug("Try starting control master by calling:\n%s" % cmd)
+        import subprocess
+        proc = subprocess.Popen(cmd, shell=True)
+        proc.communicate(input="\n")  # why the f.. this is necessary?
+
+        runner = Runner()
+        ssh_cmd = ["ssh", "-S", control_path, host_name]
+
+        for repo in repos:
+
+            if target in repos[repo].git_get_remotes():
+                # TODO: skip or raise? If raise, do it before looping, in order
+                # to fail without messing up things by partially doing the shit
+                # - what about just adding new push url to that remote?
+                # - if url(s) fit, then still try to create target?
+
+                # cmd = ["git", "remote", "get-url", "--push", remote]
+                # out, err = runner.run(cmd, cwd=handle_repo.path)
+                # if handle_remote_url_push != out.strip():
+                lgr.warning("Sibling {0} already exists. Skipping".format(target))
+                continue
+
+            # %NAME
+            REPO_NAME = repo.replace("/", "-")
+
+            # create remote repository
+            path = target_dir.replace("%NAME", REPO_NAME)
+
+            if path != '.':
+                # check if target exists, and if not --force is given,
+                # fail here
+                # TODO: Is this condition valid for != '.' only?
+                path_exists = True
+                cmd = ssh_cmd + ["ls", path]
+                try:
+                    out, err = runner.run(cmd, expect_fail=True,
+                                          expect_stderr=True)
+                except CommandError as e:
+                    if "%s: No such file or directory" % path in e.stderr:
+                        path_exists = False
+                    else:
+                        raise  # It's an unexpected failure here
+
+                if path_exists and not force:
+                    raise RuntimeError("Target directory %s already exists." %
+                                       path)
+
+                cmd = ssh_cmd + ["mkdir", "-p", path]
+                try:
+                    runner.run(cmd)
+                except CommandError as e:
+                    lgr.error("Remotely creating target directory failed at "
+                              "%s.\nError: %s" % (path, str(e)))
+                    continue
+
+            # init git repo
+            cmd = ssh_cmd + ["git", "-C", path, "init"]
```

Comment:
May be even let the GitRepo/AnnexRepo classes use a SSH() instance, so we
can just call their methods as if they were locally available ones?
But definitely let's finish the "distribution cycle" before trying to
simplify things like that.

2016-04-06 22:36 GMT+02:00 Yaroslav Halchenko <notifications@github.com>:

> In datalad/distribution/create_publication_target_sshwebserver.py
> <https://github.com/datalad/datalad/pull/388#discussion_r58777229>:
>
> > +                        raise  # It's an unexpected failure here
> > +
> > +                if path_exists and not force:
> > +                    raise RuntimeError("Target directory %s already exists." %
> > +                                       path)
> > +
> > +                cmd = ssh_cmd + ["mkdir", "-p", path]
> > +                try:
> > +                    runner.run(cmd)
> > +                except CommandError as e:
> > +                    lgr.error("Remotely creating target directory failed at "
> > +                              "%s.\nError: %s" % (path, str(e)))
> > +                    continue
> > +
> > +            # init git repo
> > +            cmd = ssh_cmd + ["git", "-C", path, "init"]
>
> may be even make use of https://github.com/paramiko/paramiko/ . Might
> simplify use of datalad on obscure OSs which lack ssh client? ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/datalad/datalad/pull/388/files/e4edaf123862ec987be3bed16903808d8e967847#r58777229>
>